af/fastpath/fastpath/core.py

```diff
@@ -350,52 +351,599 @@ def match_fingerprints(measurement):
     return matches
 
 
-@metrics.timer("score_measurement")
-def score_measurement(msm, matches):
-    """Calculate measurement scoring. Returns a summary dict
+def truekeys(d, keys):
+    """Check for values set to True in a dict"""
+    if isinstance(keys, str):
+        keys = (keys,)
+    for k in keys:
+        if d.get(k, None) != True:
+            return False
+
+    return True
+
+
+def falsekeys(d, keys):
+    """Check for values set to True in a dict"""
+    if isinstance(keys, str):
+        keys = (keys,)
+    for k in keys:
+        if d.get(k, None) != False:
+            return False
+
+    return True
+
+
+def nonekeys(d, keys):
+    """Check for values set to None in a dict"""
+    if isinstance(keys, str):
+        keys = (keys,)
+    for k in keys:
+        if d.get(k, True) != None:
+            return False
+
+    return True
+
+
+def logbug(id: int, desc: str, msm: dict):
+    """Log unexpected measurement contents, possibly due to a bug in the probe
+    The id helps locating the call to logbug()
     """
-    # Extract interesting fields
-    fields = (
-        "input",
-        "measurement_start_time",
-        "probe_asn",
-        "probe_cc",
-        "report_id",
-        "test_name",
-        "test_start_time",
+    # TODO: use assertions for unknown bugs
+    rid = msm.get("report_id", "")
+    url = "https://explorer.ooni.org/measurement/{}".format(rid) if rid else "no rid"
+    sname = msm.get("software_name", "unknown")
+    sversion = msm.get("software_version", "unknown")
+    if id > 0:
```

Comment:
What is the logic by which the `id` is getting set/passed to this function?

Reply:
I see you are always calling this with the value of `0` which should result in it going into the code path below. Why then not just check for `== 0`?

Reply:
The id is just an unique identifier for the bug so we can correlate logs and code. I could have used the line number of the caller in the core.py but that would change over time.
The description field already contains an optional, longer description and even bug number on GitHub, so the id is just a small number.

Reply:
I figured it was going to be something like that, however it's unclear to me why you chose <= 0 as for **known** bugs and > 0 for unknown ones.

Perhaps you can have a single int value for all unknowns and everything else (which will be a range) for known bugs?

I would also mention how to use the `id` parameters in `logbug` docstring.

Reply:
I'm indeed using 0 for known bugs because the description replaces the id. The negative numbers are currently unused so we can use them in future if needed.